### PR TITLE
feat: changes to add or amend claims logic (CMC-1490) (2/2)

### PR DESCRIPTION
**Pull request by devmarcy**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/CMC-1490

### Change description ###
- Created a new function in `ParticularsOfClaimValidator` that will validate for addOrAmendClaimDocuments event, as now ParticularsOfClaim document and text are optional (also accounting for backwards compatible code for particulars of claim)
- created tests in `AddOrAmendClaimDocumentsCallbackHandlerTest` 
- amended tests in `ServedDocumentFilesTest`

( https://github.com/hmcts/civil-damages-ccd-definition/pull/41 must be merged first)
**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/unspec/model/ServedDocumentFiles.java [hunk 1]
```diff
@@ -29,10 +29,18 @@
...
         }
         return errors;
     }
+
+    @JsonIgnore
+    public List<String> getErrorsAddOrAmendDocuments() {
+        List<String> errors = new ArrayList<>();
+        if (ofNullable(particularsOfClaimDocument).isPresent() && ofNullable(particularsOfClaimText).isPresent()) {
+            errors.add("More than one Particulars of claim details added");
+        }
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/model/ServedDocumentFiles.java]: can we be clever here, this is the same as line 28-31

File: src/main/java/uk/gov/hmcts/reform/unspec/validation/interfaces/ParticularsOfClaimValidator.java [hunk 2]
```diff
@@ -17,4 +17,14 @@ default CallbackResponse validateParticularsOfClaim(CallbackParams callbackParam
             .errors(servedDocumentFiles.getErrors())
             .build();
     }
+
+    default CallbackResponse validateParticularsOfClaimAddOrAmendDocuments(CallbackParams callbackParams) {
+        ServedDocumentFiles servedDocumentFiles = ofNullable(callbackParams.getCaseData().getServedDocumentFiles())
+            .orElse(ServedDocumentFiles.builder().build());
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/validation/interfaces/ParticularsOfClaimValidator.java]: can be clever here this is the same as line 13-14

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandlerTest.java [hunk 3]
```diff
@@ -68,6 +69,19 @@ void shouldReturnNoErrors_whenParticularOfClaimsFieldsAreValid() {
...
+                                                                        .particularsOfClaimText("Some string")
+                                                                        .particularsOfClaimDocument(Document.builder()
+                                                                        .build()).build()).build();
+            CallbackParams params = callbackParamsOf(caseData, MID, pageId);
+
+            var response = (AboutToStartOrSubmitCallbackResponse) handler.handle(params);
+
+            assertThat(response.getErrors()).containsOnly("More than one Particulars of claim details added");
+        }
+
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandlerTest.java:91]: should there be a test in this file for when just one document added for particulars of claim? i.e doc or text

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandler.java [hunk 4]
```diff
@@ -28,7 +28,7 @@
     protected Map<String, Callback> callbacks() {
         return Map.of(
             callbackKey(ABOUT_TO_START), this::emptyCallbackResponse,
-            callbackKey(MID, "particulars-of-claim"), this::validateParticularsOfClaim,
+            callbackKey(MID, "particulars-of-claim"), this::validateParticularsOfClaimAddOrAmendDocuments,
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandler.java]: is this the right place? I thought this story was for claimIssue event?

Comment (devmarcy) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandler.java]: this story involves changing the validation in addOrAmendClaimDocuments event

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandlerTest.java [hunk 5]
```diff
@@ -68,6 +69,32 @@ void shouldReturnNoErrors_whenParticularOfClaimsFieldsAreValid() {
...
+        @Test
+        void shouldReturnNoErrors_whenOnlyParticularOfClaimsTextSubmitted() {
+            CaseData caseData = caseDataBuilder.servedDocumentFiles(ServedDocumentFiles.builder()
+                                                                        .particularsOfClaimText("Some string").build())
+                                                                        .build();
+            CallbackParams params = callbackParamsOf(caseData, MID, pageId);
+
+            var response = (AboutToStartOrSubmitCallbackResponse) handler.handle(params);
+            System.out.println(response.getErrors());
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/user/AddOrAmendClaimDocumentsCallbackHandlerTest.java]: please remove this

**Review by dharmendrak — DISMISSED** [review]

**Review by sabahirfan — DISMISSED** [review]

**Review by dharmendrak — APPROVED** [review]

**Review by sabahirfan — APPROVED** [review]

**Comment by sabahirfan** [comment]
Closing this as this is archived repo. the work is completed in new repo. @devmarcy  feel free to re-open if you disagree.

**Pull request closed without merging**